common/src/main/java/com/sequenceiq/cloudbreak/health/HikariStateHealthIndicator.java

```diff
@@ -0,0 +1,74 @@
+package com.sequenceiq.cloudbreak.health;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.sql.DataSource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.util.ThreadUtil;
+import com.zaxxer.hikari.HikariDataSource;
+
+@Component
+public class HikariStateHealthIndicator implements HealthIndicator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HikariStateHealthIndicator.class);
+
+    @Value("${hikari.pool.readiness.down-if-pool-is-full}")
+    private boolean setReadinessProbeDownIfPoolIsFull;
+
+    @Inject
+    private List<DataSource> dataSources;
+
+    @Inject
+    private HikariDataSourcePoolMetadataExtractor hikariDataSourcePoolMetadataExtractor;
+
+    @Override
+    public Health health() {
+        boolean poolIsFull = false;
+        Map<String, Object> details = new HashMap<>();
+        try {
+            for (DataSource dataSource : dataSources) {
+                if (dataSource instanceof HikariDataSource) {
+                    HikariDataSource hikariDataSource = (HikariDataSource) dataSource;
+                    String poolName = hikariDataSource.getPoolName();
+                    DataSourcePoolMetadata poolMetadata = hikariDataSourcePoolMetadataExtractor.extract(hikariDataSource);
+                    details.put(poolName, poolMetadata);
+                    if (poolMetadata.getActive() >= poolMetadata.getMax()) {
+                        LOGGER.warn("Application has run out of open db connections. name: {}, details: {}", poolName, poolMetadata);
+                        poolIsFull = true;
+                    }
+                }
+            }
+            LOGGER.debug("HikariPool details: {}", details);
+            threadDumpIfNeeded(poolIsFull);
+        } catch (RuntimeException exception) {
+            LOGGER.info("Failed while checking the health of hikari pool!", exception);
+        }
+        return buildHealthResponse(poolIsFull, details);
+    }
+
+    private Health buildHealthResponse(boolean poolIsFull, Map<String, Object> details) {
+        Health health;
+        if (poolIsFull && setReadinessProbeDownIfPoolIsFull) {
+            health = Health.down().withDetails(details).build();
+        } else {
+            health = Health.up().withDetails(details).build();
+        }
+        return health;
+    }
+
+    private void threadDumpIfNeeded(boolean poolIsFull) {
+        if (poolIsFull) {
+            LOGGER.info("Application has run out of open db connections. Thead dump: {}", ThreadUtil.dump());
```

Comment:
Finding these logs will be very hard because they will be split by elastic search into more parts. What do you think about a solution to print each ThreadInfo to a new LOGGER line and set a requestId to the MDC context? I don't know if we set requestId and initialize MDC context at the beginning of health() call, but if not, we should initialize it.

Reply:
Good idea!

Reply:
Typo: "Thread dump"

Reply:
The dump is large, but I thought that large logs are split during shipping but eventually they will be combined. I can easily put a for loop around this, but in this case we are going to have a few hundred lines of log for every dump.

Regarding the requestid, it should be a generic solution, if we don't have a requestid, which we certainly won't have for the readiness call, then we can autogenerate one and autofill the mdc context.

Reply:
I have refactored it and added it as a new commit for review. We can squash the commits upon merge.

Reply:
Thanks, I am fine with this solution!